internal/bgp/bgpfrr/frr.go

```diff
@@ -0,0 +1,178 @@
+package bgpfrr
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"go.universe.tf/metallb/internal/bgp"
+)
+
+// As the MetalLB controller should handle messages synchronously, there should
+// no need to lock this data structure. TODO: confirm this.
+type frrState struct {
+	sessions map[string]*Session
+}
+
+var state *frrState
+
+func init() {
+	state = &frrState{
+		sessions: map[string]*Session{},
+	}
+}
+
+type Session struct {
+	myASN      uint32
+	routerID   net.IP // May be nil, meaning "derive from context"
```

Comment:
I ended up with an invalid router in my lab right off a fresh deployment:

```
/ # cat /etc/frr/frr.conf 

log stdout 
hostname 

router bgp 64512
  no bgp default ipv4-unicast
  bgp router-id <nil>

  neighbor 10.89.0.5:179 remote-as 64512
```

Reply:
Just to highlight that there are two things here that need addressing: router-id set to `<nil>` and hostname not set

```
2021-09-24T13:01:44.954061686Z b'line 3: % Command incomplete: hostname \n\n'
```